### PR TITLE
p2p. Modify peer.SendMsg which can accept any data type

### PR DESCRIPTION
**Pull request by qiaozhigang**
Modify peer.SendMsg which can accept any data type
Extract connection class from p2p.peer, supporting band meter in the future

---
This change is [Reviewable](https://reviewable.io/reviews/seeleteam/go-seele/34)

### CODE REVIEW

File: p2p/simulations/p2ptest.go [hunk 1]
```diff
@@ -19,26 +19,33 @@ import (
 type myProtocol struct {
 	p2p.Protocol
 	proto int
+	peers map[*p2p.Peer]bool
```
Comment (WTer) [inline p2p/simulations/p2ptest.go]: nice to add some comments for this var.

File: p2p/simulations/p2ptest.go [hunk 2]
```diff
@@ -19,26 +19,33 @@ import (
...
 		case peer := <-p.AddPeerCh:
-			fmt.Println("myProtocol new peer", peer)
+
+			p.peers[peer] = true
+			fmt.Println("myProtocol add new peer. peers=", len(p.peers))
 		case peer := <-p.DelPeerCh:
-			fmt.Println("myProtocol del peer", peer)
+			fmt.Println("myProtocol del peer")
+			// need del from peers
+			delete(p.peers, peer)
```
Comment (WTer) [inline p2p/simulations/p2ptest.go]: what happened when the peer already been deleted?

File: p2p/simulations/p2ptest.go [hunk 3]
```diff
@@ -75,14 +87,14 @@ func main() {
 		//slice29, _ := hexutil.HexToBytes(node29)
 		//fmt.Println(slice29)
 		nodeID29 := common.HexToAddress(node29)
-		addr29, _ := net.ResolveUDPAddr("udp4", "182.87.223.29:39009")
+		addr29, _ := net.ResolveUDPAddr("udp4", "182.87.223.29:39008")
```
Comment (WTer) [inline p2p/simulations/p2ptest.go]: where is 182.87.223.29?
and nice to define the port 39008 as constant or to be configuable

File: p2p/conn.go [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+*  @copyright defined in go-seele/LICENSE
+ */
+
+package p2p
+
+import (
+	"net"
+)
+
+// connection TODO add band meter for connection
```
Comment (WTer) [inline p2p/conn.go]: what's meaning of band meter?

File: p2p/message.go [hunk 5]
```diff
@@ -18,25 +18,26 @@ const (
 	ctlMsgPongCode       uint16 = 4
 )
 
-// Message exposed for high level layer to call
+// Message exposed for high level layer to receive
 type Message struct {
-	msgCode    uint16 // message code, defined in each protocol
-	size       uint32 // size of the paylod
```
Comment (WTer) [inline p2p/message.go]: why do you remove the size?

File: p2p/peer.go [hunk 6]
```diff
@@ -86,6 +86,12 @@ loop:
 	p.log.Info("p2p.peer.run quit. err=%s", p.err)
 }
 
+func (p *Peer) close() {
```
Comment (WTer) [inline p2p/peer.go]: do we need to add error handler for this function?

File: p2p/message.go (hunk 5 above)
Comment (qiaozhigang) [reply p2p/message.go]: As a slice, payload has size attribute

File: p2p/message.go [hunk 7]
```diff
@@ -18,25 +18,26 @@ const (
...
 	ReceivedAt time.Time
 	CurPeer    *Peer // peer that handle this message
 }
 
 // msg wrapped Message, used in p2p layer
 type msg struct {
-	Message
 	protoCode uint16
+	msgCode   uint16
+	size      uint32
```
Comment (lucusfly) [inline p2p/message.go]: you could remove this size too

File: p2p/message.go [hunk 8]
```diff
@@ -18,25 +18,26 @@ const (
...
 
-// protoHandShake handshake message for two peer to exchage base information
+// ProtoHandShake handshake message for two peer to exchage base information
 // TODO add public key or other information for encryption?
-type protoHandShake struct {
-	caps   []Cap
-	nodeID discovery.NodeID
-	nounce uint32 //
+type ProtoHandShake struct {
+	Caps   []Cap
```
Comment (lucusfly) [inline p2p/message.go]: what does Caps mean? add comments

File: p2p/peer.go [hunk 9]
```diff
@@ -140,25 +159,30 @@ func (p *Peer) handle(msgRecv *msg) error {
...
+// SendMsg called by subProtocols, message can be any data type
+func (p *Peer) SendMsg(proto *Protocol, msgCode uint16, message interface{}) error {
 	protoCode, ok := p.capMap[proto.cap().String()]
 	if !ok {
 		return errors.New("Not Found protoCode")
 	}
+	fmt.Println("SendMsg called", msgCode)
+	payload, err := common.Serialize(message)
+	if err != nil {
+		return errors.New("SendMsg yet message is not valid")
```
Comment (lucusfly) [inline p2p/peer.go]: include err message

File: p2p/message.go (hunk 7 above)
Comment (qiaozhigang) [reply p2p/message.go]: no, cannot assume the input slice has the exact size.

**Comment by lucusfly** [comment]
Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/seeleteam/go-seele/34)*

File: p2p/simulations/p2ptest.go (hunk 1 above)
Comment (qiaozhigang) [reply p2p/simulations/p2ptest.go]: p2ptest.go is for test，just show how to work with p2p.Protocol. 
Extra functions can be determined later, config-support for example

File: p2p/simulations/p2ptest.go (hunk 2 above)
Comment (qiaozhigang) [reply p2p/simulations/p2ptest.go]: Peer will send msg to DelPeerCh, subprotocol should handle it.

File: p2p/message.go (hunk 8 above)
Comment (qiaozhigang) [reply p2p/message.go]: ? same meaning with eth

File: p2p/simulations/p2ptest.go (hunk 3 above)
Comment (qiaozhigang) [reply p2p/simulations/p2ptest.go]: Should read from config

File: p2p/message.go (hunk 8 above)
Comment (lucusfly) [reply p2p/message.go]: what's the meaning in eth, could you add a comments

File: p2p/server.go [hunk 10]
```diff
@@ -188,6 +189,10 @@ func (srv *Server) scheduleTasks() {
 		if ok {
 			continue
 		}
+		/*fmt.Println("\tscheduleTasks, curNode=", node.ID)
+		for _, p := range srv.peers {
+			fmt.Println("\tpeer's node = ", p.node.ID)
+		}*/
```
Comment (lucusfly) [inline p2p/server.go]: please remove unused comments

**Pull request merged**